### PR TITLE
chore(spec-macro): Simplify calling spec conform logic

### PR DESCRIPTION
**Pull request by tangrammer**
the idea is use different validations inside any logic ... (without having several nested if/elses)

Screenshot 2022-04-22 at 12 47 46

### CODE REVIEW

File: src/main/com/kubelt/lib/config/system.cljc [hunk 1]
```diff
@@ -28,7 +28,7 @@
 (def default
   {;; These empty defaults should be overridden from the SDK init
    ;; options map.
-   :log/level nil
+   :log/level {:min-level :info}
```
Comment (crimeminister) [inline src/main/com/kubelt/lib/config/system.cljc:31]: We currently set a default log level (`:warn`) in `lib.config/log-level` that should be extracted and set as this value:

```
        log-level (get options :log/level)
        ;; Update the system configuration map before initializing the
        ;; system.
        system (-> system
                   (assoc :log/level log-level)
```

 The intent is to define all the defaults together in one place, and to treat all the values extracted from the options map uniformly: extracting them from the options map, setting up the initial values in the system map in the `init` function call as above.

File: src/main/com/kubelt/sdk/v1.cljs [hunk 2]
```diff
@@ -10,7 +10,8 @@
    [com.kubelt.lib.error :as lib.error]
    [com.kubelt.lib.init :as lib.init]
    [com.kubelt.lib.promise :refer [promise promise?]]
-   [com.kubelt.spec.config :as spec.config]))
+   [com.kubelt.spec.config :as spec.config])
+  (:require-macros [com.kubelt.spec :as kspec]))
```
Comment (crimeminister) [inline src/main/com/kubelt/sdk/v1.cljs:14]:
```suggestion
  (:require-macros
    [com.kubelt.spec :as kspec]))
```
Minor style thing for consistency with the other `require`s.

File: src/main/com/kubelt/lib/error.cljc [hunk 3]
```diff
@@ -28,7 +28,10 @@
 (defn explain
   "Return human-friendly description of why some value doesn't conform to
   the provided schema."
-  [schema value]
+  [schema value & [type-error]]
```
Comment (crimeminister) [inline src/main/com/kubelt/lib/error.cljc:31]: I am not sure I understand the intent of allowing the error type to be overridden?

File: src/main/com/kubelt/spec/config.cljc [hunk 4]
```diff
@@ -51,6 +52,17 @@
    [:p2p/multiaddr {:optional true} multiaddr]
    [:p2p/scheme {:optional true} spec.http/scheme]])
 
+(def sdk-config
```
Comment (crimeminister) [inline src/main/com/kubelt/spec/config.cljc:55]: I am not sure I understand why we want to split the config schema to make a version that supports all optional key/value pairs, and another that makes the values required?

**Review by crimeminister — APPROVED** [review]

File: src/main/com/kubelt/lib/error.cljc (hunk 3 above)
Comment (tangrammer) [reply src/main/com/kubelt/lib/error.cljc:31]: so my intent was to be more specific about error-type if provided, so something like `:kubelt.type/spec-error`
but not sure if `:com.kubelt/type` is opened to that

File: src/main/com/kubelt/spec/config.cljc (hunk 4 above)
Comment (tangrammer) [reply src/main/com/kubelt/spec/config.cljc:55]: IMO having 2 schemas is powerful! :) 
a couple of reasons:
- we are ensuring defaults are properly merged when they are needed
having every function args "speced" (exact shape of data that internal impl expects) improve not also validation but understanding of the api (internal and public ones) 🤓 
- we can mock/generate sdk values but also system ones

File: src/main/com/kubelt/spec.clj [hunk 5]
```diff
@@ -0,0 +1,9 @@
+(ns com.kubelt.spec
+  (:require
+   [malli.core :as m]
+   [com.kubelt.lib.error :as lib.error]))
+
+(defmacro conform [spec data & body]
+  `(if (m/validate ~spec ~data)
+     (do ~@body)
+     (lib.error/explain ~spec ~data :kubelt.type/spec-error)))
```
[the remark this answers is not in the document]
Comment (tangrammer) [reply src/main/com/kubelt/spec.clj:9]: your comment makes fully sense, I'll take a look to that lib and I'll be back on this topic

File: src/main/com/kubelt/lib/config/system.cljc (hunk 1 above)
Comment (tangrammer) [reply src/main/com/kubelt/lib/config/system.cljc]: solved by https://github.com/kubelt/kubelt/pull/336/commits/241de288aec23bd35c7923cab6c07973f9f1fc64

File: src/main/com/kubelt/lib/error.cljc (hunk 3 above)
Comment (tangrammer) [reply src/main/com/kubelt/lib/error.cljc:31]: I'll revert this change until we can discuss and develop better the proposal of error taxonomies and/or using anomalies lib

Comment (tangrammer) [reply src/main/com/kubelt/lib/error.cljc]: https://github.com/kubelt/kubelt/pull/336/commits/831e3cdd34970ac42a6287e9741b973f3ea58a1a

File: src/main/com/kubelt/spec.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (tangrammer) [reply src/main/com/kubelt/spec.clj]: reverting the change until further discussion on this interesting design topic :) 
https://github.com/kubelt/kubelt/pull/336/commits/831e3cdd34970ac42a6287e9741b973f3ea58a1a

**Comment by tangrammer** [comment]
merging now thus the changes will target chore/refactor-sdk-config branch,
thanks for the review!!

**Pull request merged**